### DISCUSSION TITLE
Serial read seems to be very buggy and unreliable (unless I'm using it wrong)

### DISCUSSION THREAD

**Discussion by SparkyTD** [Rahix/avr-hal#610] [Q&A] [answered]
I am trying to incorporate USB serial communication into a more complex program that does several operations between serial reads. While working on this, I have noticed a strange issue where any delay before the `serial.read()` call will cause most incoming bytes to be dropped for some reason. Here's a simple demonstration of a serial echo program:
```rust
#![no_std]
#![no_main]

use arduino_hal::delay_ms;
use arduino_hal::prelude::*;
use panic_halt as _;
use ufmt::uwriteln;

#[arduino_hal::entry]
fn main() -> ! {
    let dp = arduino_hal::Peripherals::take().unwrap();
    let pins = arduino_hal::pins!(dp);
    let mut serial = arduino_hal::default_serial!(dp, pins, 9600);

    loop {
        delay_ms(100);
        if let Ok(byte) = serial.read() {
            serial.write(byte).unwrap();
            continue;
        }
    }
}
```
This program is supposed to read all incoming bytes, and echo them back to the serial port. The issue is, that the `delay_ms(100)` causes most of the incoming bytes to be ignored.
Example input: ABCDEFGH
Output: AB

If I recreate the same program in C++, everything works as expected:
```c++
void setup() {
  Serial.begin(9600);
}

void loop() {
  delay(100);
  if (Serial.available() > 0) {
    Serial.print(String((char)Serial.read()));
  }
}
```
Example input: ABCDEFGH
Output: ABCDEFGH

This version of the program does exactly what it's supposed to do, which is echo back every character read with a 100ms delay.

I assume this is happening due to some internal hidden buffering that exists in the Arduino C++ framework which presumably isn't implemented in the Rust HAL. Is there a way to make this program function the same way it does in C++?

**Comment by Rahix** [comment]
The serial implementation here in `avr-hal` works differently than the one in Arduino C++:  We don't have any internal buffers while they do.  So if you don't read by the time each new bytes becomes available, you will loose data as you noticed.

There are two solutions:

1. Either just read fast enough and don't delay.
2. Or move the reads into the corresponding RX interrupt and write received data into a buffer.  This is essentially what the C++ implementation does under the hood as well.

**Reply by Rahix** [reply, marked as answer]
You can activate the relevant interrupt using the [`listen()`](https://rahix.github.io/avr-hal/avr_hal_generic/usart/struct.Usart.html#method.listen) method.  Check the [interrupt examples](https://github.com/Rahix/avr-hal/blob/main/examples/arduino-uno/src/bin/uno-ext-interrupt.rs) for some code example of setting up an interrupt handler.

**Reply by SparkyTD** [reply]
Awesome, thank you! This seems like exactly what I need.

**Reply by Rahix** [reply]
If you do end up writing some code like this and it is reusable enough, feel free to submit a new example showcasing this :)

**Reply by SparkyTD** [reply]
Alright, will do!
